trace_http/src/metrics.rs

```diff
@@ -0,0 +1,152 @@
+#[derive(Debug)]
+pub struct MetricsCollection {
+    is_grpc: bool,
+    metric_registry: Arc<metric::Registry>,
+    metrics: Mutex<HashMap<Option<String>, Metrics>>,
+}
+
+impl MetricsCollection {
+    pub fn new(metric_registry: Arc<metric::Registry>, is_grpc: bool) -> Self {
+        Self {
+            is_grpc,
+            metric_registry,
+            metrics: Default::default(),
+        }
+    }
+
+    /// Gets the `MetricsRecorder` for a given http request
+    pub fn recorder<B>(self: &Arc<Self>, request: &http::Request<B>) -> MetricsRecorder {
+        MetricsRecorder {
+            metrics: Arc::clone(self),
+            start_instant: Instant::now(),
+            path: Some(request.uri().path().to_string()),
+            classification: None,
+        }
+    }
+
+    fn request_metrics(&self, path: Option<String>) -> MappedMutexGuard<'_, Metrics> {
+        MutexGuard::map(self.metrics.lock(), |metrics| {
+            let (_, request_metrics) =
+                metrics.raw_entry_mut().from_key(&path).or_insert_with(|| {
+                    let attributes = match path.as_ref() {
+                        Some(path) => Attributes::from([("path", path.clone().into())]),
+                        None => Attributes::from([]),
+                    };
+
+                    let metrics =
+                        Metrics::new(self.metric_registry.as_ref(), attributes, self.is_grpc);
+                    (path, metrics)
+                });
+            request_metrics
+        })
+    }
+}
+
+/// The request metrics for a specific set of attributes (e.g. path)
+#[derive(Debug)]
+struct Metrics {
+    ok: U64Counter,
+    client_error: U64Counter,
+    server_error: U64Counter,
+    aborted: U64Counter,
+
+    duration_ok: DurationHistogram,
+    duration_client_error: DurationHistogram,
+    duration_server_error: DurationHistogram,
+}
+
+impl Metrics {
+    fn new(registry: &metric::Registry, attributes: impl Into<Attributes>, is_grpc: bool) -> Self {
+        let (counter, duration) = match is_grpc {
+            true => ("grpc_requests", "grpc_request_duration"),
+            false => ("http_requests", "http_request_duration"),
+        };
+
+        let counter: Metric<U64Counter> =
+            registry.register_metric(counter, "accumulated total requests");
+
+        let duration: Metric<DurationHistogram> =
+            registry.register_metric(duration, "distribution of request latencies");
+
+        let mut attributes = attributes.into();
+
+        attributes.insert("status", "ok");
+        let ok = counter.recorder(attributes.clone());
+        let duration_ok = duration.recorder(attributes.clone());
+
+        attributes.insert("status", "client_error");
+        let client_error = counter.recorder(attributes.clone());
+        let duration_client_error = duration.recorder(attributes.clone());
+
+        attributes.insert("status", "error");
```

Comment:
why not use `server_error` to be more explicit (and follow the pattern from `client_error`)?

```suggestion
        attributes.insert("status", "server_error");
```

Reply:
I was trying to be backwards compatible, but I think it is less confusing to be explicit - will update